### PR TITLE
WIP: Support of merge/note/pipeline events in multibranch pipelines

### PR DESCRIPTION
**Pull request by Kindrat**
Reworked trigger lookup to support merge/note/pipeline events in multibranch pipelines. There is a disgusting part - parsing data from pipeline script string to GitlabPushTrigger object - would appreciate any help in replacing this code with DSL parser.

### CODE REVIEW

**Comment by omehegan** [comment]
Let's call this a WIP, I'm not in a rush to merge any code that the author him/herself would call "disgusting" :)  Maybe we can get someone else to review this.

**Comment by Kindrat** [comment]
@omehegan thanks. Just need to know how to use DSL parser to get DTO instead of parsing string with patterns...

**Comment by omehegan** [comment]
@Kindrat what is the status of this PR? Do you need some help with review or improving it? Also I have introduced some conflicts, can you resolve them?

**Comment by Kindrat** [comment]
@omehegan that's in message just above yours

>  Just need to know how to use DSL parser to get DTO instead of parsing string with patterns...

**Comment by omehegan** [comment]
@Kindrat OK, I guess I'm not sure how you would do that. You might want to ask for help on the jenkins-dev Google group, it's likely that someone there could help.

**Comment by omehegan** [comment]
@Kindrat were you able to get any help with your questions for this PR? I was specifically suggesting you mind try asking at https://groups.google.com/forum/#!forum/jenkinsci-dev

Please let me know if you want to keep this PR open.

**Comment by Kindrat** [comment]
Was sent to read docs instead of any help... close it.

**Comment by omehegan** [comment]
Thread: https://groups.google.com/forum/#!searchin/jenkinsci-dev/oleksii%7Csort:date/jenkinsci-dev/68ETC9QthCQ/cErWHke-CQAJ

File: pom.xml [hunk 1]
```diff
@@ -164,8 +172,8 @@
     </dependency>
     <dependency>
       <groupId>org.jenkins-ci.plugins.workflow</groupId>
-      <artifactId>workflow-step-api</artifactId>
-      <version>1.15</version>
+      <artifactId>workflow-multibranch</artifactId>
```
Comment (jglick) [inline pom.xml:175]: Generally speaking, if you have to do this, you did something wrong. I do not really grasp what the goal of this PR is, so I cannot say what.

File: src/main/java/com/dabsquared/gitlabjenkins/environment/GitLabEnvironmentContributor.java [hunk 2]
```diff
@@ -18,12 +18,12 @@
...
 public class GitLabEnvironmentContributor extends EnvironmentContributor {
     @Override
-    public void buildEnvironmentFor(@Nonnull Run r, @Nonnull EnvVars envs, @Nonnull TaskListener listener) throws IOException, InterruptedException {
+    public void buildEnvironmentFor(@Nonnull Run r, @Nonnull EnvVars envs, @Nonnull TaskListener listener) {
         GitLabWebHookCause cause = null;
         if (r instanceof MatrixRun) {
             MatrixBuild parent = ((MatrixRun)r).getParentBuild();
             if (parent != null) {
-                cause = (GitLabWebHookCause) parent.getCause(GitLabWebHookCause.class);
+                cause = parent.getCause(GitLabWebHookCause.class);
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/environment/GitLabEnvironmentContributor.java:26]: Seems like unrelated code cleanups? Better left to distinct PRs.

File: src/main/java/com/dabsquared/gitlabjenkins/util/TriggerUtil.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                trigger.setJob(((WorkflowMultiBranchProject) project).getItemByBranchName(branch));
+                return trigger;
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return null;
+    }
+
+    private static GitLabPushTrigger parseTriggerFromPipeline(String script) throws AbortException {
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/util/TriggerUtil.java:87]: This is a bad idea. See discussion on the dev list, and reference to JENKINS-41929.

File: src/main/java/com/dabsquared/gitlabjenkins/webhook/build/NoteBuildAction.java [hunk 4]
```diff
@@ -34,15 +34,7 @@ public NoteBuildAction(Item project, String json, String secretToken) {
...
-        if (!(project instanceof Job<?, ?>)) {
-            throw HttpResponses.errorWithoutStack(409, "Note Hook is not supported for this project");
-        }
-        ACL.impersonate(ACL.SYSTEM, new BuildWebHookAction.TriggerNotifier(project, secretToken, Jenkins.getAuthentication()) {
-            @Override
-            protected void performOnPost(GitLabPushTrigger trigger) {
-                trigger.onPost(noteHook);
-            }
-        });
+        ACL.impersonate(ACL.SYSTEM, new BuildWebHookAction.TriggerAction<>(noteHook, project, secretToken));
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/webhook/build/NoteBuildAction.java:37]: No, never do this without a `try`-`finally` block (or the Java 7 equivalent)!

**Comment by omehegan** [comment]
@Kindrat do you want to do anything more with this?

**Comment by Kindrat** [comment]
@omehegan I still do not know how to parse Jenkinsfile so won't spend time on review and conflict fixes.
